integrations/oci/oci-base/src/main/java/io/helidon/integrations/oci/Oci.java

```diff
@@ -0,0 +1,335 @@
+        clientConfig = builder.getClientConfig();
+    }
+
+    /**
+     * The OCI create method.
+     *
+     * @param config from SE.
+     * @return using builder pattern.
+     */
+    public static Oci create(Config config) {
+        return builder().config(config).build();
+    }
+
+    /**
+     * Get the registered provider.
+     *
+     * @return AuthenticationDetailsProvider.
+     */
+    public AuthenticationDetailsProvider getProvider() {
+        return provider;
+    }
+
+    /**
+     * Get the Configured client.
+     *
+     * @return ClientConfiguration.
+     */
+    public ClientConfiguration getClientConfig() {
+        return clientConfig;
+    }
+
+    /**
+     * Access to Builder.
+     *
+     * @return The Builder.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Tenancy ID resolver.
+     *
+     * @param authenticationDetailsProvider
+     * @return TenancyIdResolver when running in the Oracle cloud
+     */
+    public String tenancyIdResolve(BasicAuthenticationDetailsProvider authenticationDetailsProvider) {
+
+        //Consul with OCI for this:
+        String metadataServiceUrl = "http://169.254.169.254/opc/v1/";
+
+        if (authenticationDetailsProvider == null) {
+            throw new OciException("Invalid Oracle Cloud Configuration");
+        }
+
+        if (authenticationDetailsProvider instanceof AuthenticationDetailsProvider) {
+            return ((AuthenticationDetailsProvider) authenticationDetailsProvider).getTenantId();
+        } else if (authenticationDetailsProvider instanceof ResourcePrincipalAuthenticationDetailsProvider) {
+            return ((ResourcePrincipalAuthenticationDetailsProvider) authenticationDetailsProvider)
+                    .getStringClaim(ResourcePrincipalAuthenticationDetailsProvider.ClaimKeys.TENANT_ID_CLAIM_KEY);
+        } else if (authenticationDetailsProvider instanceof InstancePrincipalsAuthenticationDetailsProvider) {
+            URLBasedX509CertificateSupplier urlBasedX509CertificateSupplier;
+            String tenantId;
+            try {
+                urlBasedX509CertificateSupplier = new URLBasedX509CertificateSupplier(
+                        new URL(metadataServiceUrl + "identity/cert.pem"),
+                        new URL(metadataServiceUrl + "identity/key.pem"),
+                        (char[]) null);
+                tenantId = AuthUtils.getTenantIdFromCertificate(
+                        urlBasedX509CertificateSupplier.getCertificateAndKeyPair().getCertificate()
+                );
+            } catch (MalformedURLException e) {
+                throw new OciException("Unable to retrieve tenancy ID from metadata.");
+            }
+            return tenantId;
+        }
+        return null;
+
+    }
+
+    private static final class Builder implements io.helidon.common.Builder<Oci> {
+
+        private AuthenticationDetailsProvider provider;
```

Comment:
There should be setters for these two fields, in `build` method - only create them if not explicitly configured.